exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -514,29 +517,22 @@ func getNewMachine(mp *expv1.MachinePool, infraMachine *unstructured.Unstructure
 func (r *MachinePoolReconciler) infraMachineToMachinePoolMapper(ctx context.Context, o client.Object) []ctrl.Request {
 	log := ctrl.LoggerFrom(ctx)
 
-	machinePoolList := &expv1.MachinePoolList{}
 	labels := o.GetLabels()
-	selector := map[string]string{}
-	if clusterName, ok := labels[clusterv1.ClusterNameLabel]; ok {
-		selector = map[string]string{clusterv1.ClusterNameLabel: clusterName}
-	}
-
-	if poolNameHash, ok := labels[clusterv1.MachinePoolNameLabel]; ok {
-		if err := r.Client.List(ctx, machinePoolList, client.InNamespace(o.GetNamespace()), client.MatchingLabels(selector)); err != nil {
-			log.Error(err, "failed to get MachinePool for InfraMachine")
+	_, machinePoolOwned := labels[clusterv1.MachinePoolNameLabel]
+	if machinePoolOwned {
+		machinePool, err := utilexp.GetMachinePoolByLabels(ctx, r.Client, o.GetNamespace(), labels)
+		if err != nil {
+			log.Error(err, "failed to get MachinePool for InfraMachine", "infraMachine", klog.KObj(o), "labels", labels)
```

Comment:
Continuing the conversion about GetMachinePoolByLabels returning an error if the machine is not labelled, even though the mapping function doesn't return the error, it would generate a log here.

Could you test this PR with a non-machinepool cluster? If we do end up calling this mapping function on all machines (not 100% sure about that), this would be highly undesirable. We'd have to filter it somehow.

Reply:
Sure, let me give that a try and get back to you.

Reply:
@mdbooth Sorry for the late reply, but I tested this out with #8842 and didn't see the error in the logs. That being said, I'm fine with removing the return error `errors.Errorf("labels missing required key `%s`", clusterv1.MachinePoolNameLabel)` and just have it return nil instead. WDYT?